File: services/appview/src/cursor.ts
type SortMode = 'newest' | 'vote_date' | 'topic' | 'popularity';

export interface CursorPayload {
  sort: SortMode;
  p: string; // primary key (date, topic, popularity) as string
  r: string; // rkey
}


export function encodeCursor(payload: CursorPayload): string {
  return Buffer.from(JSON.stringify(payload), 'utf8').toString('base64url');
}

export function decodeCursor(cursor: string): CursorPayload {
  return JSON.parse(Buffer.from(cursor, 'base64url').toString('utf8'));
}

File: services/appview/cursor.py
import json
import base64
from typing import TypedDict


class CursorPayload(TypedDict):
    sort: str  # 'newest' | 'vote_date' | 'topic' | 'popularity'
    p: str     # primary key (date, topic, popularity) as string
    r: str     # rkey


def encode_cursor(payload: CursorPayload) -> str:
    json_str = json.dumps(payload)
    return base64.urlsafe_b64encode(json_str.encode('utf-8')).decode('utf-8')


def decode_cursor(cursor: str) -> CursorPayload:
    decoded = base64.urlsafe_b64decode(cursor.encode('utf-8')).decode('utf-8')
    return json.loads(decoded)
